cmd/config-encrypted.go

```diff
@@ -217,89 +249,155 @@ func migrateIAMConfigsEtcdToEncrypted(ctx context.Context, client *etcd.Client)
-	}
 
-	err := saveConfig(GlobalContext, objAPI, backendEncryptedFile, backendEncryptedMigrationIncomplete)
-	if err != nil {
-		return err
+		if !res.IsTruncated {
+			break
+		}
+		marker = res.NextMarker
 	}
+	return nil
+}
+
+func migrateEncryptedConfig(objAPI ObjectLayer, key, oldKey config.MasterKey, oldCred auth.Credentials) error {
+	logger.Info("Attempting (re)encryption of all config, IAM users and policies on MinIO backend")
 
 	var marker string
 	for {
-		res, err := objAPI.ListObjects(GlobalContext, minioMetaBucket,
-			minioConfigPrefix, marker, "", maxObjectList)
+		res, err := objAPI.ListObjects(GlobalContext, minioMetaBucket, minioConfigPrefix, marker, "", maxObjectList)
 		if err != nil {
 			return err
 		}
 		for _, obj := range res.Objects {
-			var (
-				cdata    []byte
-				cencdata []byte
-			)
-
-			cdata, err = readConfig(GlobalContext, objAPI, obj.Name)
+			var data, plaintext []byte
+			data, err = readConfig(GlobalContext, objAPI, obj.Name)
 			if err != nil {
 				return err
 			}
 
-			var data []byte
-			// Is rotating of creds requested?
-			if activeCredOld.IsValid() {
-				data, err = decryptData(cdata, activeCredOld, globalActiveCred)
-				if err != nil {
-					if err == madmin.ErrMaliciousData {
-						return config.ErrInvalidRotatingCredentialsBackendEncrypted(nil)
-					}
-					return err
+			// The data might be:
+			//  - encrypted with the new/current master key (incomplete migration)
+			//  - encrypted with the old master key (secret key rotation)
+			//  - encrypted with the old or current secret key using the "legacy" format (migration)
+			// We first try the current master key, then the old master key and finally the legacy
+			// decryption scheme. We do this in this order, since trying a decryption with a master key
+			// is much cheaper then trying the legacy scheme decryption. (HMAC vs. Argon2id)
+			// So, the migration from "legacy" to "new" format is slightly slower - but a secret key rotation
+			// is significantly faster. Since the format migration is a one-time opeation (but key rotation is not)
+			// this seems justifable.
+			plaintext, err = key.DecryptBytes(data)
+			if err == sio.NotAuthentic {
+				if oldKey != nil {
+					plaintext, err = oldKey.DecryptBytes(data)
 				}
-			} else {
-				data = cdata
-			}
-
-			if !utf8.Valid(data) {
-				_, err = decryptData(data, globalActiveCred)
-				if err == nil {
-					// Config is already encrypted with right keys
-					continue
+				if err == sio.NotAuthentic {
+					plaintext, err = decryptData(data, oldCred, globalActiveCred) // "legacy" scheme
 				}
-				return errors.New("config data not in plain-text form or encrypted")
 			}
-
-			cencdata, err = madmin.EncryptData(globalActiveCred.String(), data)
 			if err != nil {
 				return err
 			}
 
-			if err = saveConfig(GlobalContext, objAPI, obj.Name, cencdata); err != nil {
+			data, err = key.EncryptBytes(plaintext)
```

Comment:
For future compatibility, could we add a version identifier so we can identify files saved with exactly this configuration.

Whether it is in the object metadata or a byte before the encrypted payload isn't super important to me, but we should have something other than just a pure payload so we can uniquely identify it. We have enough hacks and workarounds.

Reply:
That cannot be done with e.g. `file.v1` scheme - in a secure way - since you have to take into account for malicious modification. So, such a versioning does not allow a general simplification.
Instead see: https://github.com/minio/minio/pull/9757#discussion_r435215587

Reply:
What do you mean? As stated it makes no sense, and I guess you are misunderstanding.

For the file stored on the backend we need to know which serialization version did the encryption in case it needs to be changed. Otherwise you have to do even more hacks to determine the format of the following content.

This will not change malicious modification angle.

Reply:
> For the file stored on the backend we need to know which serialization version did the encryption in case it needs to be changed.

This is a) already addressed by https://github.com/minio/minio/pull/9757#discussion_r435215587 and b) *very unlikely* due to proofs. 

> For future compatibility, could we add a version identifier so we can identify files saved with exactly this configuration.

How do you plan to **reliably** determine the configuration used to produce the file without decrypting the file?

Reply:
How do you plan to reliably determine the configuration used to produce the file without decrypting the file?

@aead Ah, ok , this is where I was unclear. I was unable to link up the connection. If that is already handled, then we are 👍